design/pipeline-caching.md

```diff
@@ -0,0 +1,154 @@
+
+For experiment #3 we used a dedup file transfer mechanism built into Azure Artifacts (used behind the scenes by Universal Packages and Pipeline Artifacts). As part of the experiment we packed the node_modules folder into a tar file (without compression) which resulted in a 150MB file. We were able to transfer that file in about 19 seconds and then unpacking took about another 28 seconds - resulting in the same performance as a Yarn install on a clean cache (no improvement).
+
+In experiment #4 we eliminated the tar file and allowed the dedup file transfer to directly place files on disk. This process took 22 seconds end to end which roughly halves the Yarn installation time.
+
+In experiment #5 we instead cached the Yarn package cache directory instead of the local ```node_modules``` folder. The download of the packages into the Yarn cache directory took 22 seconds similar to experiment #4, however Yarn installation added 22 seconds for linking which meant it was roughly the same as a clean install. My comparison, running Yarn install with experiment #4 is close to a no-op for Yarn.
+
+In order to explore the archive file scenario a little bit more we did experiment #6 where in addition to creating a ```.tar``` archive, we also added compression to the file (```.tgz```). Due to file size reductions the transfer time was reduced to about 7 seconds, but decompression and placing files using ```tar``` took 35 seconds.
+
+Based on the above results it would appear that using our dedup file transfer mechanism with direct file placement would be the most performant option, however from internal experimentation at Microsoft we've observed that some larger code bases with hundreds of thousands of files in the ```node_modules``` directory do better with ```.tgz``` so we believe to service the breadth of caching requirements for Azure Pipelines users that we'll need to support a number of ways of placing files.
+
+Other more elaborate caching mechanisms may also be available such as virtualized file systems with prefetch which may provide significant performance boosts particularly in sparse file access scenarios but introduce other issues which would need to make that approach strictly opt-in.
+
+The caching strategy would default to  dedup file transfer mechanism but allow an option to override the caching strategy with another approach, for example"
+
+Selecting the cache strategy would simply be a matter of specifying a strategy on the task inputs (for both the restore and save tasks, restore shown below):
+
+```yaml
+steps:
+- task: RestoreCache@0
+  inputs:
+    strategy: dedup
+    key: |
+      package.json
+      yarn.lock
+    path: node_modules
+```
+
+### YAML syntax
+
+The example task usage above uses the explicit task references. We also want to provide a streamlined syntax for caching, applied to the example above.
+
+```yaml
+steps:
+- restoreCache: yarn.lock
+- script: yarn
+- saveCache:
+  key: yarn.lock
```

Comment:
Do we have to say "key" here or can we mirror the `restoreCache` model of just putting the key after `saveCache`?

Reply:
It will be possible (and sometimes necessary) to interleave multiple cache restore and save operations. For example:

```yaml
steps:
- restoreCache: yarn.lock
  path: node_modules
- script: npm install
- script: ./node_modules/bin/grunt build
- restoreCache: Service.csproj
  path: packages
- saveCache: yarn.lock
  path: node_modules
  path: node_modules
- script: dotnet build && dotnet test
- saveCache: Service.csproj
  path: packages
```

Its a bit contrived but you can probably see how sometimes people will interleave caches (especially if caches are related and interdependent).

That said - maybe there is a simple scenario that we can optimize here. I recall when we caught up in NC that you mentioned that there could be a way of doing something here?

Reply:
Not sure I follow. I was asking a syntactic question about whether `key` has to live as a separate input vs if it could just go after `saveCache`. Given that you did it in your example, I think your answer is yes? 😛 

Regarding the `path` stuff -- you can't have duplicate `path` keys. You have two options: (1) make `path` take either a string or an array of strings or (2) take only a string, and in the task, expect a newline-delimited list.

Reply:
What would happen if in V1 of the task we converted the type to string[]? The YAML ends up being the same doesn't it?

Reply:
We have the ability to say "take a scalar or an array". But if you choose just one in the schema, then your YAML is limited to that one. Does that make sense?

```yaml
scalar: foo
array:
- bar
- baz
```